pkg/auto/history/airqualitysensor.go

```diff
@@ -67,3 +68,48 @@ func (a *automation) collectAirQualityChanges(ctx context.Context, source config
 		a.logger.Warn("collection aborted", zap.Error(err))
 	}
 }
+
+func (a *automation) sampleAirQualityChanges(ctx context.Context, source config.Source, payloads chan<- []byte) {
+	var client traits.AirQualitySensorApiClient
+	if err := a.clients.Client(&client); err != nil {
+		a.logger.Error("sampling aborted", zap.Error(err))
+		return
+	}
+
+	t := time.Now()
+
+	var prev *traits.AirQuality
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Until(source.Sample.Schedule.Next(t))):
+			resp, err := client.GetAirQuality(ctx, &traits.GetAirQualityRequest{Name: source.Name, ReadMask: source.ReadMask.PB()})
+
+			t = time.Now()
+			if err != nil {
+				a.logger.Warn("sample aborted", zap.Error(err))
+				continue
+			}
+
+			if proto.Equal(prev, resp) {
+				continue
+			}
+
+			prev = proto.Clone(resp).(*traits.AirQuality)
+
+			payload, err := proto.Marshal(proto.Clone(resp).(*traits.AirQuality))
+			if err != nil {
+				a.logger.Warn("sample aborted", zap.Error(err))
+				continue
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case payloads <- payload:
+			}
```

Comment:
Almost all this code is a duplicate, with the pattern being repeated between all the traits can you get rid of the duplicated code please.

Reply:
Also see the comment below about the code in factory.go